export.go

```diff
@@ -396,7 +395,7 @@ func (conn *Conn) export(methods map[string]reflect.Value, path ObjectPath, ifac
 
 	// If this is the first handler for this path, make a new map to hold all
 	// handlers for this path.
-	if !h.PathExists(path) {
+	if _, ok := h.PathExists(path); !ok {
```

Comment:
Better get a write lock here then and keep it until the end of the function - this also removes the need for checking the other `PathExists` in L408.

Reply:
The main problem I can see with that is what lock should be taken, I would argue that it is not a good design to call h.Lock() from this context and if we choose to do that we need to remove the lock call in h.AddObject too, making it asymmetric to h.RemoveObject. To me it is better to let defaultHandler and its functions handle its own locks.

Reply:
But the way this is implemented right now introduces a race condition. If you have two `export` calls running at the same time for the same path, it could happen that:
1. Call 1 reaches export.go:398. There's no object on that path yet, so `ok` is false. After `GetExportedObject` returns, the lock is not held.
2. Call 2 goes through the whole function, working on the newly created entry in `h.objects`.
3. Call 1 is scheduled again and reaches export.go:399, overwriting the entry in `h.objects` and thus discarding the effect that Call 2 had.